### PR TITLE
DNS Plugin: Domain Offensive

### PR DESCRIPTION
**Pull request by Armitxes**
Thanks for your work. Thought I share my DNS Plugin with you that I use for my hoster [Domain Offensive](https://www.do.de/). Hope I didn't oversee anything but if so, feel free to modify it.

# How To Use the Domain Offensive DNS Plugin

This plugin works against the [Domain Offensive](https://www.do.de/) DNS provider. It is assumed that you have already setup an account and created the DNS zone(s) you will be working against.

## Setup

We need to retrieve an secret API token for the account that will be used to update DNS records. Further information can ba found at [the (german) developer docs.](https://www.do.de/wiki/LetsEncrypt_-_Entwickler).

## Using the Plugin

Your personal API token is specified using the `DomOffToken` or `DomOffTokenInsecure` parameter.

### Windows and/or PS 6.2+ only (secure string)
```powershell
$pArgs = @{
    DomOffToken = $secureString
}
New-PACertificate example.com -DnsPlugin DomainOffensive -PluginArgs $pArgs
```

### Any OS (default string)
```powershell
$pArgs = @{
    DomOffTokenInsecure = '1md6xRcUCTrB58kbpwAH'
}
New-PACertificate example.com -DnsPlugin DomainOffensive -PluginArgs $pArgs
```

### CODE REVIEW

**Comment by rmbolger** [comment]
Hey @Armitxes, thanks for the plugin! This is the first provider I've seen that has an API specifically for Let's Encrypt.

The only problem I see at the moment is that the parameter blocks in the functions aren't quite right. The `$RecordName` and `$TxtValue` params should remain as the first and second in both the add and remove functions. Both also need to still have the `$ExtraParams` param in there at the end as well. This allows people to use multiple different plugins for names in a single cert.

I would also rename the `$Token` param to something more specific to Domain Offensive like `$DomOffToken`. Each plugin's parameters need to be unique among all plugins.

If you want to make things even better, take a look at how some of the other plugins offer a "secure" and "insecure" version of things like secret tokens. This allows them to have their secret values encrypted when they're saved on the local drive.

File: Posh-ACME/DnsPlugins/DomainOffensive-Readme.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Your personal API token is specified using the generated `Token` parameter.
+The fully qualified name for the TXT record can be specified using the `RecordName` parameter.
+The value for the TXT record can be specified using the `TxtValue` parameter.
+
+```powershell
+$pArgs = @{
+    Token = '1md6xRcUCTrB58kbpwAH',
+    Domain = '_acme-challenge.site1.example.com',
+    TxtValue = 'OVxwaDm7MgN1IRG0eSivJMlepO9CL4X8vKo6Tcns'
```
Comment (rmbolger) [inline Posh-ACME/DnsPlugins/DomainOffensive-Readme.md:19]: These shouldn't be necessary to pass as plugin parameters. They're automatically passed by the module to the plugin.

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 [hunk 2]
```diff
@@ -0,0 +1,103 @@
+function Add-DnsTxtDomainOffensive {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory,Position=0)]
+        [string]$Token,
+        [Parameter(Mandatory,Position=1)]
+        [string]$RecordName,
+        [Parameter(Mandatory,Position=2)]
+        [string]$TxtValue
```
Comment (rmbolger) [inline Posh-ACME/DnsPlugins/DomainOffensive.ps1:9]: These need to be positional params 0 and 1. And the `$ExtraParams` one needs to be added back at the end of the param block.

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    #>
+}
+
+function Remove-DnsTxtDomainOffensive {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory,Position=0)]
+        [string]$Token,
+        [Parameter(Mandatory,Position=1)]
+        [string]$RecordName
```
Comment (rmbolger) [inline Posh-ACME/DnsPlugins/DomainOffensive.ps1:52]: This param block should be identical to the Add one even if you don't need all of the values.

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    .PARAMETER Token
+        Token provided by Domain Offensive.
+
+    .PARAMETER Domain
+        The fully qualified name of the TXT record to be removed.
+
+    .EXAMPLE
+        Add-DnsTxtDomainOffensive '1md6xRcUCTrB58kbpwAH' '_acme-challenge.site1.example.com'
+
+        Adds a TXT record for the specified site with the specified value using the account associated with the given token.
```
Comment (rmbolger) [inline Posh-ACME/DnsPlugins/DomainOffensive.ps1:79]: I think you forgot to update this example after copying it from the Add function.

File: Posh-ACME/DnsPlugins/DomainOffensive-Readme.md (hunk 1 above)
Comment (Armitxes) [reply Posh-ACME/DnsPlugins/DomainOffensive-Readme.md]: Good to know. Made the mistake to pass the parameters myself.
Fixed.

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 (hunk 2 above)
Comment (Armitxes) [reply Posh-ACME/DnsPlugins/DomainOffensive.ps1]: Fixed order and added the `$ExtraParams`

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 (hunk 3 above)
Comment (Armitxes) [reply Posh-ACME/DnsPlugins/DomainOffensive.ps1]: Done

File: Posh-ACME/DnsPlugins/DomainOffensive.ps1 (hunk 4 above)
Comment (Armitxes) [reply Posh-ACME/DnsPlugins/DomainOffensive.ps1]: Yep, did a little Oopsie there.. fixed.
Thanks for reviewing :)

**Comment by Armitxes** [comment]
@rmbolger Thanks for your review. I think I covered all of your suggestions. Tested with secure string and it worked aswell.

Let me know if there is anything else to do.

**Comment by rmbolger** [comment]
Awesome. Thanks for the updates. Made some tiny additional tweaks with the help stuff. But otherwise, this is good to go.

**Pull request merged**